### PR TITLE
WIP - Customer authentication

### PR DESCRIPTION
**Pull request by agrzelewska**
Signup doesn/t work, unfortunately. HELP, please! :)

Web browser shows :"Corrupted Content Error" and terminal: 
Started POST "/customers" for 127.0.0.1 at 2018-06-03 14:27:14 +0200
Processing by CustomersController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"IqjzEjN5Bn+deoAlUYyrfcf2w1jgmq7dlmyMvYvL2sw+BT1xtd7NBjjz1CLV8LT8hqFtWgp0PLcHfJ1Nas+I0Q==", "customer"=>{"name"=>"Kate", "email"=>"kate@example.com", "password_digest"=>"[FILTERED]"}, "commit"=>"Sign up"}
Unpermitted parameter: :password_digest
   (1.0ms)  BEGIN
  ↳ app/controllers/customers_controller.rb:8
  Customer Exists (2.3ms)  SELECT  1 AS one FROM "customers" WHERE "customers"."email" = $1 LIMIT $2  [["email", "kate@example.com"], ["LIMIT", 1]]
  ↳ app/controllers/customers_controller.rb:8
   (1.6ms)  ROLLBACK
  ↳ app/controllers/customers_controller.rb:8
Redirected to http://localhost:3000signup
Completed 302 Found in 35ms (ActiveRecord: 4.8ms)

### CODE REVIEW

**Comment by zulixro** [comment]
Look for an error message. In this case it's `Unpermitted parameter: :password_digest`. Try to find out what's going on with that. Hint: look up "permited parameters"

**Comment by agrzelewska** [comment]
I've seen that. I've tried to find solution but I think that customer_params are correct (when I've changed :password to :password_digest the error was the same)

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+	helper_method :current_customer
+
+	def current_customer
+		@current_customer ||= Customer.find(session[:customer_id]) if session[:customer_id]
+	end
+
+	def require_customer
+		redirect_to '/login' unless current_customer
```
Comment (gpks) [inline app/controllers/application_controller.rb]: in ruby we use 2 spaces indentations :)

**Comment by gpks** [comment]
I checked - if you change `:password` into `:password_digest` in permitted params, as @zulixro wrote, this error is not appearing again, but there is another problem then...

**Comment by agrzelewska** [comment]
I did it :). I've changed password into password_digest in validation in models/concerns/customers.rb . And now it signs up into db :)

**Comment by agrzelewska** [comment]
I did some commits - I think it's working now, at least better :)

File: app/controllers/customers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class CustomersController < ApplicationController
+  def new
+   @customer = Customer.new
+  end
+
+  def create
+    @customer = Customer.new(customer_params)
+    if @customer.save
```
Comment (gpks) [inline app/controllers/customers_controller.rb]: please fix indentations here: if you don't know how, I will show you today :)

File: app/controllers/customers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	    redirect_to @customer
+	    # After signing up customer goes to the customers#show page:
+	    #redirect_to @customer
+	  else
+	    #render 'new'
+      redirect_to '/signup'
+	  end
+  end
+
+  def show
```
Comment (gpks) [inline app/controllers/customers_controller.rb]: do we need this method?

File: app/controllers/customers_controller.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  end
+
+  private
+	  def customer_params
```
Comment (gpks) [inline app/controllers/customers_controller.rb]: indentations: method should be at the same level as `private`

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    @customer = Customer.find_by_email(params[:session][:email])
+	if @customer && @customer.authenticate(params[:session][:password])
```
Comment (gpks) [inline app/controllers/sessions_controller.rb]: indentations

File: app/views/customers/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<h1>Sign up</h1>
+
+<%= form_for(@customer) do |f| %>
+    <%= f.label :name %><br>
```
Comment (gpks) [inline app/views/customers/new.html.erb:4]: 👍

File: app/views/customers/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    <%= f.text_field :name %><br>
+    <%= f.label :email %><br>
+    <%= f.email_field :email %><br>
+    <%= f.label :password %><br>
+    <%= f.password_field :password %><br>
+    <%= f.submit "Sign up" %>
+<% end %>
```
Comment (gpks) [inline app/views/customers/new.html.erb]: here's why github is showing you this red sign: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

**Review by gpks — CHANGES_REQUESTED** [review]
I have some remarks, but overall: amazing work!!!!! 👏

File: app/views/customers/show.html.erb [hunk 8]
```diff
@@ -0,0 +1 @@
+dadada
```
Comment (gpks) [inline app/views/customers/show.html.erb]: newline please :)

File: app/views/sessions/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<%= form_for(:session, url: login_path) do |f| %>
+  <%= f.email_field :email, :placeholder => "Email" %>
+  <%= f.password_field :password, :placeholder => "Password" %>
+  <%= f.submit "Log in" %>
+<% end %>
```
Comment (gpks) [inline app/views/sessions/new.html.erb]: newline please

File: test/models/customer_test.rb [hunk 10]
```diff
@@ -1,7 +1,54 @@
...
+  def setup
+    @customer = Customer.new(name: "Example Customer", email: "customer@example.com", password: "foobar")
+  end
+
+  test "should be valid" do
+  	assert @customer.valid?
+  end
+
+  test "name should be present" do
+  	@customer.name = "    "
```
Comment (gpks) [inline test/models/customer_test.rb]: indentations 😩

**Review by gpks — APPROVED** [review]

**Pull request merged**